### PR TITLE
Avoid loading round state for dropped messages

### LINKED ISSUE #208: `instance.roundState(round)` is a (minor) DoS vector
IIRC, the plan to avoid storing arbitrary messages was to:

1. Store at most one "quality" value from every miner.
2. ignore all messages from future rounds that lack justification from a prior round (again, at most once per miner).

However, `roundState` implicitly allocates a round object before we even look at the message. It's possible for a miner to repeatedly send the _same_ quality message but with different rounds, forcing us to allocate a bunch of round states.

This will _probably_ get fixed when we implement the above but... it's a subtle implicit allocation that's easily overlooked, so I wanted to write this down so we don't forget.

### PR DESCRIPTION
**Pull request by anorth**
Closes #208

### CODE REVIEW

File: gpbft/gpbft.go [hunk 1]
```diff
@@ -262,7 +267,7 @@ func (i *instance) ReceiveMany(msgs []*GMessage) error {
 	// Received each message and remember which rounds were received.
 	roundsReceived := map[uint64]struct{}{}
 	for _, msg := range msgs {
-		if err := i.receiveOne(msg); err != nil {
+		if _, err := i.receiveOne(msg); err != nil {
```
Comment (anorth) [inline gpbft/gpbft.go:270]: This path is reached only once per instance, so I don't think it's worth the check here. But I can accumulate a stateChanged here too if requested.

File: gpbft/gpbft.go [hunk 2]
```diff
@@ -294,90 +299,90 @@ func (i *instance) Describe() string {
...
+			msgRound.committed.ReceiveJustification(msg.Vote.Value, msg.Justification)
 		}
 	case DECIDE_PHASE:
 		i.decision.Receive(msg.Sender, msg.Vote.Value, msg.Signature)
 		if i.phase != DECIDE_PHASE {
 			i.skipToDecide(msg.Vote.Value, msg.Justification)
 		}
-		if err := i.tryDecide(); err != nil {
-			return fmt.Errorf("failed to decide: %w", err)
-		}
```
Comment (anorth) [inline gpbft/gpbft.go:361]: This is redundant after implementation of skipToDecide, immediately above. `tryDecide()` will be invoked by `tryCurrentPhase()`  at the end.

File: gpbft/gpbft.go [hunk 3]
```diff
@@ -391,22 +396,21 @@ func (i *instance) postReceive(roundsReceived ...uint64) {
...
 func (i *instance) shouldSkipToRound(round uint64, state *roundState) (ECChain, *Justification, bool) {
-
 	// Check if the given round is ahead of current round and this instance is not in
 	// DECIDE phase.
 	if round <= i.round || i.phase == DECIDE_PHASE {
 		return nil, nil, false
 	}
+	if !state.prepared.ReceivedFromWeakQuorum() {
+		return nil, nil, false
+	}
```
Comment (anorth) [inline gpbft/gpbft.go:406]: Moved this cheaper and harder-to-meet check to happen first

File: gpbft/gpbft.go (hunk 1 above)
Comment (Stebalien) [reply gpbft/gpbft.go:270]: We're limited to `maxLookaheadRounds * powerLookback` anyways (except for justified messages). So I'd say this is fine.

Comment (Stebalien) [reply gpbft/gpbft.go:270]: Although... it should be trivial to just not add this round to `roundsReceived`, right?

**Review by Stebalien — APPROVED** [review]

File: gpbft/gpbft.go (hunk 1 above)
Comment (anorth) [reply gpbft/gpbft.go:270]: Yes,done

**Pull request merged**